src/Components/Server/src/Circuits/RemoteJSDataStream.cs

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits
+{
+    internal sealed class RemoteJSDataStream : Stream
+    {
+        private readonly RemoteJSRuntime _runtime;
+        private readonly long _streamId;
+        private readonly long _totalLength;
+        private readonly CancellationToken _streamCancellationToken;
+        private readonly Stream _pipeReaderStream;
+        private readonly Pipe _pipe;
+        private long _bytesRead;
+
+        public static async Task<bool> ReceiveData(RemoteJSRuntime runtime, long streamId, byte[] chunk, string error)
+        {
+            if (!runtime.RemoteJSDataStreamInstances.TryGetValue(streamId, out var instance))
+            {
+                // There is no data stream with the given identifier. It may have already been disposed.
+                // We notify JS that the stream has been cancelled/disposed.
+                return false;
+            }
+
+            return await instance.ReceiveData(chunk, error);
+        }
+
+        public static async ValueTask<RemoteJSDataStream> CreateRemoteJSDataStreamAsync(
+            RemoteJSRuntime runtime,
+            IJSDataReference jsDataReference,
+            long totalLength,
+            long maxBufferSize,
+            long maximumIncomingBytes,
+            CancellationToken cancellationToken = default)
+        {
+            var streamId = runtime.RemoteJSDataStreamNextInstanceId++;
+            var remoteJSDataStream = new RemoteJSDataStream(runtime, streamId, totalLength, maxBufferSize, cancellationToken);
+            await runtime.InvokeVoidAsync("Blazor._internal.sendJSDataStream", jsDataReference, streamId, maximumIncomingBytes);
```

Comment:
> that we could use for sending acks to the client when we consume data instead of using the invoke logic we currently have.

A potential concern with this is the additional overhead of adding another call to send the ack. With the `invoke` logic, the ack call also contains a data payload. Additionally, going the invoke route allows us to ensure we maintain interactivity on the JS side (ex. cancel button, other actions, can be interleaved before we send in the next batch of chunks). I believe that control should be left on the JS side in the interest of separation of concerns. Moving the ack logic here could limit the interactivity on JS.

> This might also allow us to adaptatively control the window size (if we detect that we constantly run out of data when reading, it might mean we might be able to go faster).

If this specifically is of interest we can add an additional endpoint which allows the server to adjust `numChunksUntilNextAck` on the client. However, I believe this should be decoupled from the actual ack as this would likely be adjusted far less frequently than our acks.

Reply:
I think I'm not expressing myself correctly.

Reply:
I've added a bigger comment above, I get a more clear picture of what we are doing right now, but we need to add lots of comments to explain why we are doing what we are doing and the assumptions that we are making.

Reply:
Agreed that trying to clarify the sequence would be valuable. AFAIK the current mechanism with `invoke`/`send` already meets our goals well while keeping the appropriate level of control on the server. When thinking about this, we should separate these two concerns:

 * Controlling the buffering and backpressure. This is what `Pipe` manages for us and is a security concern because it relates to consumption of server resources.
 * Controlling the tradeoff between throughput and interactivity. This is what the `invoke`/`send` mechanism achieves. It's not a security concern - a badly-behaved can only harm their own interactivity/bandwidth tradeoff but can't consume extra server resources.

I think these are largely independent concerns.

Reply:
Looks like there aren't any direct action items here, safe to resolve?